PRINCIPLES.md

```diff
@@ -0,0 +1,186 @@
+
+2. [**The principle of immutable configuration versions**](#tag)
+
+    _Desired State_ is stored in a way that supports versioning, immutability of versions, and retains a complete version history.
+    We call systems that store desired state in this way _State Stores_.
+
+3. [**The principle of continuous state reconciliation**](#tag)
+
+    Software agents continuously, and automatically, compare a system's _Actual State_ to its _Desired State_.
+    If the actual and desired states differ, automated actions are immediately attempted to reconcile them.
+    These differences could be due to the actual state drifting from the desired state, or the desired state changing intentionally.
+
+4. [**The principle of operations through declaration**](#tag)
+
+    When wishing to operate on a software system, a human or software agent will not interact with the running system and modify it directly.
+    Instead, the agent will create a new declarative version of the desired state in the state store.
+
+## Introduction
+
+The software systems that we manage vary widely; from battery powered embedded systems driven by microcontrollers, to globally distributed applications with millions of users. 
+Most systems are composed of subsystems themselves made of software.
+It is impossible to comprehensively outline all the specific practices for managing such a variety of systems.
+
+However, despite the myriad differences, several important principles emerge that simplify the task of reliably managing and operating _all_ software systems at scale. GitOps attempts to capture some of these principles in a coherent framework.
+The GitOps principles can be applied to managing entire software systems or applied only to parts of larger systems. 
+
+## Scope
+
+GitOps concerns the verifiable behaviour of computer systems and their interfaces. 
+Specifically, GitOps is not about human processes, and is not intended as a model for judging human organisational designs and operational practices.
+
+The GitOps principles are to be used as guiding principles in the development of modern software and system operations. They do not form a concrete specification. 
+
+The GitOps principles are a direction, not a destination. They should be applied pragmatically. For example, whilst they can be applied strictly to an entire software systems, they can also be applied loosely to selected parts of larger systems as part of a progressive adoption.
+
+## The GitOps Principles
+
+### 1. Declarative configuration
+
+<div style="border: 1px solid black; padding:5px; font-weight:bold">
+A system managed by GitOps must have its <em>Desired State</em> expressed declaritively as data. 
+ This declarative data must be in a format writable and readable by both humans and software.
+</div>
+
+<!--
+We've gone back and forth to try to find a really good single-word term for what we mean by "immutable, versioned, declarations of the desired system state".
```

Comment:
i think we need to pick this apart a bit to discuss the declaration of state vs the fulfillment of that declaration. Can we clarify that GitOps is primarily a fulfillment of state declaration of the system? that is, its a mechanism and has nothing to do with the declaration itself?

Reply:
I think this is why 'configuration' falls short - it reads as the declaration rather than the fulfillment of such

Reply:
gitops is about the auditability of a declared state, how do you come to consensus across one or more actors that the declared state is correct, and how do you fulfill the declaration of that state.

If we become too dogmatic about the format of that declaration than we risk underminig the applicability of GitOps.

Systems should be able to be closed loop, or human in the loop, to the applciation and preference of the operator.

Reply:
>  GitOps is primarily a fulfillment of state declaration of the system? that is, its a mechanism and has nothing to do with the declaration itself?

Strong disagree. The declaration of the system itself is of key importance and should meet certain criteria to enable the rest of the GitOps loop. The fulfilment of the declaration is the concern of "Continuous state reconciliation" and somewhat "Operations through declaration". 

> gitops is about the auditability of a declared state, 
I agree. I think of this as an observability requirement and tend to say that "the current state of a running system must be observable" to mean the same thing. This is, in my opinion, required for state reconciliation. I didn't add a note there explicitly though, as that section was unfinished.

> how do you come to consensus across one or more actors that the declared state is correct, 
Unless I misunderstand, that's explicitly out-of-scope as it deals with human agents coming to an agreement, which is essentially a human process. I personally have some strong opinions here, but I don't think they belong in a definition of GitOps.